contracts/ERC677BridgeToken.sol

```diff
@@ -104,5 +107,50 @@ contract ERC677BridgeToken is
         require(token.transfer(_to, balance));
     }
 
+    /**
+     * @param _signature bytes The signature, issued by the owner.
+     * @param _to address The address which you want to transfer to.
+     * @param _value uint256 The amount of tokens to be transferred.
+     * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
+     * @param _nonce uint256 Presigned transaction number. Should be unique, per user.
+     */
+    function transferPreSigned(bytes _signature, address _to, uint256 _value, uint256 _fee, uint256 _nonce) public returns (bool) {
+        require(_to != address(0), "Invalid _to address");
+
+        bytes32 hashedParams = getTransferPreSignedHash(address(this), _to, _value, _fee, _nonce);
+        address from = Message.recover(hashedParams, _signature);
+        require(from != address(0), "Invalid from address recovered");
+        bytes32 hashedTx = keccak256(abi.encodePacked(from, hashedParams));
+        require(hashedTxs[hashedTx] == false, "Transaction hash was already used");
+
+        balances[from] = balances[from].sub(_value).sub(_fee);
+        balances[_to] = balances[_to].add(_value);
+        balances[msg.sender] = balances[msg.sender].add(_fee);
+        hashedTxs[hashedTx] = true;
+        noncePerAddress[from] = noncePerAddress[from].add(1);
+
+        emit Transfer(from, _to, _value);
+        emit Transfer(from, msg.sender, _fee);
+        emit TransferPreSigned(from, _to, msg.sender, _value, _fee);
+        return true;
+    }
 
+    /**
+     * @param _token address The address of the token.
+     * @param _to address The address which you want to transfer to.
+     * @param _value uint256 The amount of tokens to be transferred.
+     * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
+     * @param _nonce uint256 Presigned transaction number.
+     */
```

Comment:
I thought about the `nonce` issue and maybe the mapping I added is redundant.
A client can use timestamp in that field, because the only demand of it is to be unique.
Replay is still prevented because it's part of the signature.

@bejavu @leonprou WDYT?

Reply:
true, if not you should check that the `_nonce == noncePerAddress[from]` in line 123.
timestamp is good.

Reply:
on second thought maybe do need the nonce per user because there can be a really rare edge case where two different clients send exactly the same values for `address _token, address _to, uint256 _value, uint256 _fee, uint256 _timestamp` so the signature will be exactly the same and one of them may fail.
So I added the missing check.